src/openassetio-core/src/errors/exceptionMessages.cpp

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2023 The Foundry Visionmongers Ltd
+
+#include "exceptionMessages.hpp"
+#include <fmt/core.h>
+
+namespace openassetio {
+inline namespace OPENASSETIO_CORE_ABI_VERSION {
+namespace errors {
+
+Str errorCodeName(BatchElementError::ErrorCode code) {
+  switch (code) {
+    case BatchElementError::ErrorCode::kUnknown:
+      return "unknown";
+    case BatchElementError::ErrorCode::kInvalidEntityReference:
+      return "invalidEntityReference";
+    case BatchElementError::ErrorCode::kMalformedEntityReference:
+      return "malformedEntityReference";
+    case BatchElementError::ErrorCode::kEntityAccessError:
+      return "entityAccessError";
+    case BatchElementError::ErrorCode::kEntityResolutionError:
+      return "entityResolutionError";
+    case BatchElementError::ErrorCode::kInvalidPreflightHint:
+      return "invalidPreflightHint";
+    case BatchElementError::ErrorCode::kInvalidTraitSet:
+      return "invalidTraitSet";
+  }
+
+  return "Unknown ErrorCode";
+}
+
+std::string createBatchElementExceptionMessage(const BatchElementError &err, size_t index,
+                                               std::optional<EntityReference> entityReference,
+                                               std::optional<access::Access> access) {
+  /*
+   * BatchElementException messages consist of five parts.
+   * 1. The name of the error code.
+   * 2. If existing, the message inside the BatchElementError.
+   * 3. The index that the batch error relates to.
+   * 4. If existing, the access.
+   * 5. If existing, the entity reference.
+   *
+   * Ends up looking something like : "entityAccessError: Could not
+   * access Entity [index=2] [access=read] [entity=bal:///entityRef]"
+   */
+  return fmt::format(
+      "{}{}{}{}{}", fmt::format("{}:", errorCodeName(err.code)),
+      err.message.empty() ? "" : fmt::format(" {}", err.message),
+      fmt::format(" [index={}]", std::to_string(index)),
+      (access.has_value() ? fmt::format(" [access={}]", access::kAccessNames[access.value()])
+                          : ""),
+      (entityReference.has_value()
+           ? fmt::format(" [entity={}]", entityReference.value().toString())
+           : ""));
+}
```

Comment:
[consider] Factoring this out from nested `fmt` calls for readability - its quite hard to grok.

Reply:
Factored out the substrings, I think it makes the construction clearer.

https://github.com/OpenAssetIO/OpenAssetIO/pull/1109/commits/8760330c5c96feb127885c4703c227c54b18bd4c

Alas the hard wrapping limit forces inconsistency and messiness (imo) here, looks very clean when the ternerys can all get a line to themself, even prior to factoring them out.

Reply:
Def an improvement, thanks. 👍  If the inconsistency is bugging you, what about something along the lines of the following (not sure if it's what @feltech was alluding to)?

```cpp
std::string result = fmt::format("{}:", errorCodeName(err.code));
if( !err.message.empty() ) {
 result += fmt::format(" {}", err.message);
}

result += fmt::format(" [index={}]", index);
if( access.has_value() ) {
    result += fmt::format(" [access={}]", access::kAccessNames[access.value()]);
}
if( entityReference.has_value() ) {
    result += fmt::format(" [entity={}]", entityReference.value().toString())
}

return result;
```

Ironicly, it kind of feels like stringstream could be easier to read here 🙃

Reply:
SS would still have the inconsistent wrapping, that's just to do with line length.
I don't think the new version is unreadable at all tbh, and it's even better now that we've YAGNI's the optionals. Resolving this.